include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp

```diff
@@ -0,0 +1,332 @@
+     *
+     * \details
+     *
+     * Resizes the entire trace matrix storing the best trace path and the horizontal trace column.
+     * Note the trace matrix requires the number of columns and rows to be one bigger than the size of sequence1,
+     * respectively sequence2 for the initialisation of the matrix.
+     * Reallocation happens only if the new column size exceeds the current capacity of the underlying trace matrix.
+     *
+     * ### Complexity
+     *
+     * In worst case `column_count` times `row_count` memory is allocated.
+     *
+     * ### Exception
+     *
+     * Basic exception guarantee. Might throw std::bad_alloc on resizing the internal matrices.
+     */
+    template <std::integral column_index_t, std::integral row_index_t>
+    void resize(column_index_type<column_index_t> const column_count,
+                row_index_type<row_index_t> const row_count)
+    {
+        this->column_count = column_count.get();
+        this->row_count = row_count.get();
+        complete_matrix.resize(number_rows{this->row_count}, number_cols{this->column_count});
+        horizontal_column.resize(this->row_count);
+        vertical_column = views::repeat_n(trace_t{}, this->row_count);
+    }
+
+    /*!\name Iterators
+     * \{
+     */
+    //!\brief Returns the iterator pointing to the first column.
+    iterator begin()
+    {
+        return iterator{*this, 0u};
+    }
+
+    //!\brief This score matrix is not const-iterable.
+    iterator begin() const = delete;
+
+    //!\brief Returns the iterator pointing behind the last column.
+    iterator end()
+    {
+        return iterator{*this, column_count};
+    }
+
+    //!\brief This score matrix is not const-iterable.
+    iterator end() const = delete;
+    //!\}
+};
+
+/*!\brief Trace matrix iterator for the pairwise alignment using the full trace matrix.
+ * \implements std::input_iterator
+ *
+ * \details
+ *
+ * Implements a counted iterator to keep track of the current column within the matrix. When dereferenced, the
+ * iterator returns a view over the allocated memory of the respective columns. The returned view zips
+ * the three columns into a single range.
+ */
+template <typename trace_t>
+//!\cond
+    requires std::same_as<trace_t, trace_directions>
+//!\endcond
+class trace_matrix_full<trace_t>::iterator
+{
+private:
+    //!\brief A lightweight representation of a single column from the complete matrix.
+    using single_trace_column_type = std::span<trace_t>;
+    //!\brief The type of the zipped score column.
+    using matrix_column_type = decltype(views::zip(std::declval<single_trace_column_type>(),
+                                                   std::declval<physical_column_t &>(),
+                                                   std::declval<virtual_column_t &>()));
+    //!\brief The column type as value type.
+    using matrix_column_value_t = std::vector<std::ranges::range_value_t<matrix_column_type>>;
```

Comment:
Just hypothetical speaking, do we need this? Does a user know how to use the value type? What are the guarantees of the value type? Does it really require a deep copy?

Maybe we should also turn-around the semantics, that the view is the reference (because it behaves much closer to a &) and that we need a special value type.

We have some places now that use views as reference type, and we might want to generalise it.

I have something like this in mind:

```c++
template <std::views::view view_t>
struct basic_view_iterator_value_type : std::vector<std::ranges::range_value_t<view_t>>
{
    using reference = view_t;
    using value_type = std::vector<std::ranges::range_value_t<view_t>>;

    basic_view_iterator_value_type(view_t view)
    { std::ranges::copy(column, std::cpp20::back_inserter(target)); }
}
```

(We force with this, that the common type is the value_type, not sure if we want this)

or

```c++
template <std::views::view view_t>
struct basic_view_iterator_reference_type : view_t
{

    using reference = view_t;
    using value_type = std::vector<std::ranges::range_value_t<view_t>>;

    constexpr operator value_type() const /*memo: const/non-const may need to be considered separately */
    {
        value_type target{};
        std::ranges::copy(*this, std::cpp20::back_inserter(target));
        return target;
    }
}
```

(We force with this, that the common type is the reference_type, not sure if we want this)

Reply:
I think in both cases the common type is the value type and also correct in this case. I mean no matter how you put it, you always convert the view into a vector.
The issue with the to_simd_view and the proposed solution with the reference proxy was, what was also hammering in my head. I solved it now like this, but I honestly prefer the special value_type with a constructor from the view more.
Simply, because in most cases I don't use the value type but the reference type and would like to avoid the indirection with the proxy.

> Just hypothetical speaking, do we need this? Does a user know how to use the value type? What are the guarantees of the value type? Does it really require a deep copy?

Basically no! The class itself is no user class and only used inside of the algorithm. So we could basically ignore that value type and reference type are not the same. However, if someone does use it and expect a correct implementation then:
```cpp
iter_value_type v = *it;
```
is weird, since it basically means self-assignment. Because the returned view itself is, as you mentioned it, also just like `&` to some memory.